pkg/asset/machines/azure/azuremachines.go

```diff
@@ -144,6 +145,34 @@ func GenerateMachines(clusterID, resourceGroup, subscriptionID string, in *Machi
 		}
 	}
 
+	defaultDiag := &capz.Diagnostics{
+		Boot: &capz.BootDiagnostics{
+			StorageAccountType: capz.ManagedDiagnosticsStorage,
+		},
+	}
+
+	if in.Platform.DefaultMachinePlatform != nil && in.Platform.DefaultMachinePlatform.BootDiagnostics != nil {
+		defaultDiag.Boot.StorageAccountType = in.Platform.DefaultMachinePlatform.BootDiagnostics.Type
+		if saURI := bootDiagStorageURIBuilder(in.Platform.DefaultMachinePlatform.BootDiagnostics, in.Platform.CloudName); saURI != "" {
+			defaultDiag.Boot.UserManaged = &capz.UserManagedBootDiagnostics{
+				StorageAccountURI: saURI,
+			}
+		}
+	}
+
+	controlPlaneDiag := defaultDiag
+	if mpool.BootDiagnostics != nil {
+		controlPlaneDiag.Boot.StorageAccountType = mpool.BootDiagnostics.Type
+		if saURI := bootDiagStorageURIBuilder(mpool.BootDiagnostics, in.Platform.CloudName); saURI != "" {
+			controlPlaneDiag.Boot.UserManaged = &capz.UserManagedBootDiagnostics{
+				StorageAccountURI: saURI,
+			}
+		}
+	}
+	if controlPlaneDiag.Boot.StorageAccountType != capz.UserManagedDiagnosticsStorage {
+		controlPlaneDiag.Boot.UserManaged = nil
+	}
```

Comment:
nit: this smells a bit to me. Rather than set `controlPlaneDiag := defaultDiag` and then unset the values like this, maybe initialize an empty `controlPlaneDiag` and if its nil after the control plane machineset check set `controlPlaneDiag = defaultDiag`. NBD

Reply:
I get that. You're unsetting a problematic field because that fixes a problem pointed out by QE. It smells to me to unset the field; it seems more intuitive to not set the problematic field in the first place.

Reply:
Ok I think it's the naming of the variable that's the problem. The value for defaultdiag is set to something yeah but the value at this point gets overriden by whatever's in the default machine. If the users provide a default machine platform boot diagnostics as "UserManaged" and then set the control plane boot diag to "Managed" then there's an object that gets populeted for default one and then gets set to managed in the control plane setting. This causes the UserManaged object to be populated for a "Managed" setting.

Reply:
You mean the naming of the variable is causing confusion? I'm not sure I follow. The problem is what I was trying to describe in my first comment. `controlPlaneDiag := defaultDiag` instantiates the whole `BootDiagnostics` struct but then if the control plane machine pool is set, only individual fields are updated; so there can be this discrepancy.

Reply:
Fixed it.